src/indexer/utils/bitcoin/index.ts

```diff
@@ -0,0 +1,255 @@
+/*
+The Licensed Work is (c) 2023 Sygma
+SPDX-License-Identifier: LGPL-3.0-only
+*/
+import FeeRepository from "indexer/repository/fee"
+import { sha256 } from "ethers"
+import { BigNumber } from "@ethersproject/bignumber"
+import { ObjectId } from "mongodb"
+import { TransferStatus } from "@prisma/client"
+import { RPCClient } from "rpc-bitcoin"
+import { BitcoinTypeTransfer, Block, DecodedDeposit, DecodedExecution, Transaction } from "../../../indexer/services/bitcoinIndexer/bitcoinTypes"
+import { BitcoinResource, Domain } from "../../../indexer/config"
+import ExecutionRepository from "../../repository/execution"
+import TransferRepository from "../../repository/transfer"
+import { logger } from "../../../utils/logger"
+import DepositRepository from "../../repository/deposit"
+import CoinMarketCapService from "../../../indexer/services/coinmarketcap/coinmarketcap.service"
+
+const WitnessV1Taproot = "witness_v1_taproot"
+const OP_RETURN = "nulldata"
+
+export async function saveEvents(
+  client: RPCClient,
+  block: Block,
+  domain: Domain,
+  executionRepository: ExecutionRepository,
+  transferRepository: TransferRepository,
+  depositRepository: DepositRepository,
+  feeRepository: FeeRepository,
+  coinMakerCapService: CoinMarketCapService,
+): Promise<void> {
+  for (const tx of block.tx) {
+    const depositData = decodeDepositEvent(tx, domain)
+    if (depositData) {
+      await saveDeposit(
+        tx.txid,
+        block.height,
+        block.mediantime,
+        domain.id,
+        depositData,
+        transferRepository,
+        depositRepository,
+        feeRepository,
+        coinMakerCapService,
+      )
+      continue
+    }
+    const executionData = await decodeExecutionEvent(tx, domain.resources as BitcoinResource[], client)
+    if (executionData) {
+      await saveProposalExecution(tx.txid, block.height, block.mediantime, domain.id, executionData, transferRepository, executionRepository)
+    }
+  }
+}
+
+function decodeDepositEvent(tx: Transaction, domain: Domain): DecodedDeposit | undefined {
```

Comment:
I think it is more intuitive to throw errors for unexpected behaviours instead of returning undefined and then we don't what happened. Also, would remove the need to have decoding and then undefined checks in `saveEvents`
Probably decoding errors will need to be caught and handled appropriately in `saveEvents` and logged as `warn` as we most likely don't want to try decoding that again, but at least we will know what happened.

Reply:
Hm, I might've made a bad approach for the whole workflow. But what this is essentially doing is checking if the output address is the relayer address (`resourceFound`) and if it is, it parses the data and stores the transfer in the database. If it isn't, it returns `undefined`. So it'll return undefined for every transaction that isn't actually a Sygma deposit.